extract/tarball.go

```diff
@@ -67,7 +72,13 @@ func ExtractTarball(reader io.Reader, filename, directory string, logger types.L
 				return fmt.Errorf("failed to create output file '%s': %w", path, err)
 			}
 
-			if _, err := io.Copy(output, tarReader); err != nil {
+			// By default, an os.File implements the io.ReaderFrom interface.
+			// As a result, CopyBuffer will attempt to use the output.ReadFrom method to perform
+			// the requested copy, which ends up calling the unbuffered io.Copy function & performing
+			// a large number of allocations.
+			// In order to force CopyBuffer to actually utilize the given buffer, we have to ensure
+			// output does not implement the io.ReaderFrom interface.
+			if _, err := io.CopyBuffer(onlyWriter{output}, tarReader, buf); err != nil {
```

Comment:
This is fun. It looks like `ReadFrom` will use `copy_file_range` on kernels >= 5.3, which would be way more efficient, but falls back to `io.Copy` otherwise, like you mentioned.